### DISCUSSION TITLE
Is there a way to call a function whenever a worker starts up?

### DISCUSSION THREAD

**Discussion by adamghill** [django-q2/django-q2#90] [Ideas]
I am using django-q2 to process some ML models. It works great, but every time a new cluster starts up, the first task for each worker calls a heavy function (which then gets cached). 

I was hoping to have a signal that fires when each worker gets created. Or I could create my own management command that gets called after `manage.py qcluster` which could use `Stat` to get all of the workers, create a task for each worker, and send a task to each worker specifically. From looking through the code, I don't think that's possible, though.

Am I missing a way I could achieve this now? If any of these approaches make sense (or if there is another approach that would work better) I'm willing to send a PR. I just wanted to check before going through the effort. Thanks, I really love all the work you've done on `django-q2`!

**Comment by adamghill** [comment]
What I've done for now is add some code to one of my app's `Config` `ready` method and that seems to run when the workers get instantiated.

**Comment by GDay** [comment]
Interesting use case.

Maybe this can be resolved by changing only one line of code:

When django-q2 runs a task, each worker has their own process. The process can be given a name (it currently uses [the default](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process.name) which is not guaranteed to be unique).

If we give that a uuid when creating the process, then you could cache items prefixing that uuid to use specifically with that worker.

For example:
```
from multiprocessing import current_process

process_name = current_process().name
result = cache.get(f"{process_name}_something")

if result is None:
    result = run_function_that_needs_to_be_cached()
    cache.set(f"{process_name}_something", result)
```

That would allow you to only run the task once and save the result and then skip it the next time (get it from cache). Would that work for you?

> What I've done for now is add some code to one of my app's Config ready method and that seems to run when the workers get instantiated.

I am not entirely sure about your setup, but wouldn't that then also run when you start the normal web server? But maybe that's intended behavior.

**Reply by adamghill** [reply]
>That would allow you to only run the task once and save the result and then skip it the next time (get it from cache). Would that work for you?

Yes! I think that would work.

>wouldn't that then also run when you start the normal web server

Yeah, it's not ideal, but it was a trade-off I was willing to make when I was trying to figure something out. Maybe I could set an environment variable and check for it, but I haven't gotten around to trying that.

**Reply by GDay** [reply]
I have just created this PR: https://github.com/django-q2/django-q2/pull/91. Feel free to take a look, I think this help with your issue. I will merge it in the next few days.

**Reply by GDay** [reply]
@adamghill That one has been merged now. The logging looks a little ugly, but I will redo those later. I have also merged a second PR that added a `post_spawn` signal for workers: https://django-q2.readthedocs.io/en/master/signals.html#after-spawning-a-worker-process. You might be able to use that too.

**Reply by adamghill** [reply]
This looks great. Thank you for adding this feature, I will try it out soon!
